src/profiler.ts

```diff
@@ -13,57 +13,251 @@
+      this.profileTypes.push(ProfileTypes.Wall);
+      this.timeProfiler =
+          new TimeProfiler(this.config.timeSamplingIntervalMicros);
+    }
   }
 
-  // Begins collection and uploading of profiles.
-  // If profiling fails or another problem is encountered, the returned promise
-  // will be rejected.
-  // TODO: implement
-  // TODO: explain failure situations.
+  /**
+   * Starts and endless loop to poll Stackdriver Profiler server for
+   * instructions, and collects and uploads profiles as requested.
+   * If there is a problem when collecting a profile or uploading a profile to
+   * Stackdriver Profiler, this problem will be logged at the debug level.
+   * If there is a problem polling Stackdriver Profiler for instructions
+   * on the type of profile created, an error will be thrown.
+   */
   async start(): Promise<void> {
-    return Promise.reject(
-        new Error('start() is unimplemented for ProfileAgent'));
+    return this.pollProfilerService();
+  }
+
+
+  /**
+   * Endlessly polls the profiler server for instructions, and collects and
+   * uploads profiles as requested.
+   *
+   * The profiler server will be polled for instructions at most once every
+   * minTimeBetweenProfilesMillis.
+   */
+  async pollProfilerService(): Promise<void> {
+    const startCreateMillis = Date.now();
+    const prof = await this.createProfile();
+    await this.profileAndUpload(prof);
+    const endCreateMillis = Date.now();
+
+    // Schedule the next profile.
+    const delayMillis = Math.max(
+        0,
+        this.config.minTimeBetweenProfilesMillis -
+            (endCreateMillis - startCreateMillis));
+
+    setTimeout(this.pollProfilerService.bind(this), delayMillis).unref();
+  }
+
+  /**
+   * Talks to Stackdriver Profiler server, which hangs until server indicates
+   * job should be profiled.
+   *
+   * If any problem is encountered, the problem will be logged and
+   * createProfile() will be retried.
+   *
+   * TODO: implement backoff and retry when error encountered. createProfile()
+   * should be retried at time response indicates this request should be retried
+   * or with exponential backoff (up to one hour) if the response does not
+   * indicate when to retry this request. Once this is implemented, an error
+   * will be thrown only if the error indicates one definitely should not
+   * retry createProfile.
+   *
+   * Public to allow for testing.
+   */
+  async createProfile(): Promise<RequestProfile> {
+    const reqBody = {
+      deployment: {
+        projectId: this.config.projectId,
+        target: this.config.serviceContext.service,
+        labels: {zone: this.config.zone, instance: this.config.instance},
+      },
+      profileType: this.profileTypes,
+    };
+    const options = {
+      method: 'POST',
+      uri: '/profiles',
+      body: reqBody,
+      json: true,
+    };
+    try {
+      const [body, response] = await this.request(options);
+      return body;
+    } catch (err) {
```

Comment:
Note that there are two types of errors possible here. Network errors (which will be thrown) and HTTP errors, which will be indicated by a non-200 HTTP statusCode in the response. The code as written, ignores HTTP errors, and assumes all network errors are ignorable.

Some more questions you need to ask: are there a subset of errors that are fatal enough that you should terminate (e.g. a ENOTFOUND thrown, or a HTTP 404 response), or if the right thing to do is to always ignore such errors and keep retrying the same operation (which are likely to fail the same way).